common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,198 @@
+    TrieMap[String, mutable.ListBuffer[ActiveSpan]]()
+  private val contextMap: mutable.Map[String, SpanContext] = TrieMap[String, SpanContext]()
+
+  var enabled = false;
+
+  def apply(serviceName: String): Unit = {
+    configureTracer(serviceName)
+  }
+
+  /**
+   * Start a Trace for given service.
+   *
+   * @param transactionId transactionId to which this Trace belongs.
+   * @return TracedRequest which provides details about current service being traced.
+   */
+  def startTrace(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+    if (enabled) {
+      var activeSpan: Option[ActiveSpan] = None
+      spanMap.get(transactionId.meta.id) match {
+        case Some(spanList) => {
+          //create a child trace
+          activeSpan = Some(
+            GlobalTracer
+              .get()
+              .buildSpan(logMarker.action)
+              .withTag("transactionId", transactionId.meta.id)
+              .asChildOf(spanList.last)
+              .startActive())
+        }
+        case None => {
+          contextMap.get(transactionId.meta.id) match {
+            case Some(context) => {
+              //create child trace if we have a tracing context
+              activeSpan = Some(GlobalTracer.get().buildSpan(logMarker.action).asChildOf(context).startActive())
+            }
+            case None => {
+              activeSpan = Some(GlobalTracer.get().buildSpan(logMarker.action).ignoreActiveSpan().startActive())
+            }
+          }
+          //initialize list for this transactionId
+          val list: mutable.ListBuffer[ActiveSpan] = mutable.ListBuffer()
+          spanMap.put(transactionId.meta.id, list)
+        }
+      }
+
+      //add active span to list
+      if (activeSpan.isDefined)
+        spanMap.get(transactionId.meta.id).map(_.+=:(activeSpan.get))
+    }
+  }
+
+  /**
+   * Finish a Trace associated with given transactionId.
+   *
+   * @param transactionId
+   */
+  def finish(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {
+    if (enabled)
+      clear(transactionId)
+  }
+
+  /**
+   * Register error
+   *
+   * @param transactionId
+   */
+  def error(transactionId: TransactionId): Unit = {
+    if (enabled)
+      clear(transactionId)
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = {
+    var contextMap: Option[Map[String, String]] = None
+    if (enabled) {
+      spanMap.get(transactionId.meta.id) match {
+        case Some(spanList) => {
+          var map: java.util.Map[String, String] = new java.util.HashMap()
+          //inject latest span context in map
+          GlobalTracer.get().inject(spanList.last.context(), Format.Builtin.TEXT_MAP, new TextMapInjectAdapter(map))
+          contextMap = Some(scala.collection.JavaConverters.mapAsScalaMapConverter(map).asScala.toMap)
```

Comment:
Instead of this you can use [JavaConverters implicits][1]

```scala
import scala.collection.JavaConverters._

contextMap = Some(map.asScala.toMap)
```

[1]: https://alvinalexander.com/scala/how-to-go-from-java-collections-convert-in-scala-interact